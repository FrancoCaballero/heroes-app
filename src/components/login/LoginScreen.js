import { useNavigate } from 'react-router-dom'
import { useContext } from 'react'
import { AuthContext } from '../../auth/authContext'
import { useForm } from '../../hooks/useForm'

import { types } from '../../types/types'

export const LoginScreen = () => {
  const navigate = useNavigate()
  const [{ username }, handleInputChange] = useForm({ username: '' })
  const { dispatch } = useContext(AuthContext)

  const handleLogin = (e) => {
    e.preventDefault()

    const action = {
      type: types.login,
      payload: {
        name: username
      }
    }
    dispatch(action)

    const lastPath = localStorage.getItem('lastPath') || '/marvel'

    navigate(lastPath, { replace: true })
  }

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />
      <form onSubmit={handleLogin}>
        <div className="col-6">

          <input className="form-control" name="username" type="text" value={username} onChange={handleInputChange} autoComplete="off"/>
          <button className="btn btn-primary mt-3">
            Login
          </button>
        </div>
      </form>

    </div>
  )
}
