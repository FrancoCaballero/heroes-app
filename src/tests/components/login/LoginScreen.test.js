import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { AuthContext } from '../../../auth/authContext'
import { LoginScreen } from '../../../components/login/LoginScreen'
import { types } from '../../../types/types'

const mockDispatch = jest.fn()
const mockNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}))

describe('pruebas en LoginScreen', () => {
  const wrapper = mount(
    <AuthContext.Provider value={{ user: { logged: false }, dispatch: mockDispatch }}>
      <MemoryRouter>
        <LoginScreen />
      </MemoryRouter>
    </AuthContext.Provider>
  )

  test('debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('debe realizar el dispatch y la navegaciÃ³n', () => {
    wrapper.find('input').simulate('change', { target: { name: 'username', value: 'Pedro' } })
    wrapper.find('form').simulate('submit', { preventDefault: () => {} })

    expect(mockDispatch).toHaveBeenCalledWith({
      type: types.login,
      payload: { name: 'Pedro' }
    })

    expect(mockNavigate).toHaveBeenCalledWith('/marvel', { replace: true })

    localStorage.setItem('lastPath', '/dc')
    wrapper.find('form').simulate('submit', { preventDefault: () => {} })

    expect(mockNavigate).toHaveBeenCalledWith('/dc', { replace: true })
  })
})
