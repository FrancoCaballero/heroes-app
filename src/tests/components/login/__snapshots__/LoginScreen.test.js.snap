// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en LoginScreen debe mostrarse correctamente 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div
      className="col-6"
    >
      <input
        autoComplete="off"
        className="form-control"
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        className="btn btn-primary mt-3"
      >
        Login
      </button>
    </div>
  </form>
</div>
`;
